Mathlib/NumberTheory/LSeries/Nonvanishing.lean

```diff
@@ -0,0 +1,241 @@
+/-
+Copyright (c) 2024 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.Analysis.SpecialFunctions.Complex.LogBounds
+import Mathlib.NumberTheory.LSeries.QuadraticNonvanishing
+
+/-!
+# The L-function of a Dirichlet character does not vanish on Re(s) ≥ 1
+
+The main result in this file is `DirichletCharacter.Lfunction_ne_zero_of_one_le_re`:
+if `χ` is a Dirichlet character and `s.re ≥ 1` and either `χ` is nontrivial or `s ≠ 1`,
+then the L-function of `χ` does not vanish at `s`.
+
+As a consequence, we have the corresponding statement for the Riemann ζ function:
+`riemannZeta_ne_zero_of_one_le_re`.
+
+These results are prerequisites for the **Prime Number Theorem** and
+**Dirichlet's Theorem** on primes in arithmetic progressions.
+-/
+
+section nonvanishing
+
+open Complex
+
+-- This is the key positivity lemma that is used to show that the L-function
+-- of a Dirichlet character `χ` does not vanish for `s.re ≥ 1` (unless `χ^2 = 1` and `s = 1`).
+private lemma re_log_comb_nonneg {a : ℝ} (ha₀ : 0 ≤ a) (ha₁ : a < 1) {z : ℂ} (hz : ‖z‖ = 1) :
+      0 ≤ 3 * (-log (1 - a)).re + 4 * (-log (1 - a * z)).re + (-log (1 - a * z ^ 2)).re := by
+  have hac₀ : ‖(a : ℂ)‖ < 1 := by
+    simp only [norm_eq_abs, abs_ofReal, _root_.abs_of_nonneg ha₀, ha₁]
+  have hac₁ : ‖a * z‖ < 1 := by rwa [norm_mul, hz, mul_one]
+  have hac₂ : ‖a * z ^ 2‖ < 1 := by rwa [norm_mul, norm_pow, hz, one_pow, mul_one]
+  have H₀ := (hasSum_re <| hasSum_taylorSeries_neg_log hac₀).mul_left 3
+  have H₁ := (hasSum_re <| hasSum_taylorSeries_neg_log hac₁).mul_left 4
+  have H₂ := hasSum_re <| hasSum_taylorSeries_neg_log hac₂
+  rw [← ((H₀.add H₁).add H₂).tsum_eq]; clear H₀ H₁ H₂
+  refine tsum_nonneg fun n ↦ ?_
+  simp only [← ofReal_pow, div_natCast_re, ofReal_re, mul_pow, mul_re, ofReal_im, zero_mul,
+    sub_zero]
+  rcases n.eq_zero_or_pos with rfl | hn
+  · simp only [pow_zero, CharP.cast_eq_zero, div_zero, mul_zero, one_re, mul_one, add_zero,
+     le_refl]
+  · simp only [← mul_div_assoc, ← add_div]
+    refine div_nonneg ?_ n.cast_nonneg
+    rw [← pow_mul, pow_mul', sq, mul_re, ← sq, ← sq, ← sq_abs_sub_sq_re, ← norm_eq_abs, norm_pow,
+      hz]
+    calc
+     0 ≤ 2 * a ^ n * ((z ^ n).re + 1) ^ 2 := by positivity
+      _ = _  := by ring
```

Comment:
alternative (not clear if it's actually better):
```suggestion
    convert (show 0 ≤ 2 * a ^ n * ((z ^ n).re + 1) ^ 2 by positivity) using 1
    ring
```

Reply:
It saves a line is certainly not worse.